raiden-ts/src/transport/epics/helpers.ts

```diff
@@ -64,101 +65,87 @@ export function getRoom$(matrix: MatrixClient, roomIdOrAlias: string): Observabl
-      filter(
-        ([{ presences }, { caps }]) =>
-          !!(
-            presences[address]?.payload.available &&
-            getCap(caps, Capabilities.TO_DEVICE) &&
-            getCap(presences[address].payload.caps, Capabilities.TO_DEVICE)
+  const rtcChannel$ =
+    content.msgtype === 'm.text' ? latest$.pipe(pluckDistinct('rtc', address)) : of(null);
+  const presence$ = latest$.pipe(pluckDistinct('presences', address));
+  const roomId$ = latest$.pipe(
+    map(({ state }) => state.transport.rooms?.[address]?.[0]),
+    distinctUntilChanged(),
+  );
+  const via$ = combineLatest([rtcChannel$, presence$, roomId$]).pipe(
+    withLatestFrom(config$),
+    map(([[rtcChannel, presence, roomId], { caps }]) => {
+      if (rtcChannel?.readyState === 'open') return rtcChannel;
+      else if (
+        getCap(caps, Capabilities.TO_DEVICE) &&
+        getCap(presence?.payload.caps, Capabilities.TO_DEVICE)
+      )
+        return presence!.payload.userId;
+      else if (roomId) return roomId;
+    }),
+    distinctUntilChanged(),
+  );
+  return via$.pipe(
+    switchMap((via) => {
+      if (!via) return EMPTY;
+      else if (typeof via !== 'string')
+        return defer(async () => {
+          via.send(content.body);
+          return via.label; // via RTC channel, complete immediately
+        });
+      else if (via.startsWith('@'))
+        return matrix$.pipe(
+          mergeMap(async (matrix) => matrix.sendToDevice(type, { [via]: { '*': content } })),
+          mapTo(via), // via toDevice message
+          // complete only when peer is online
+          delayWhen(
+            constant(
+              presence$.pipe(
+                filter(
+                  (presence) => presence?.payload.userId === via && presence.payload.available,
+                ),
+              ),
+            ),
           ),
-      ),
-      pluck(0, 'presences', address, 'payload', 'userId'),
-    ),
-    waitMember$(matrix, address, { latest$ }),
-  ).pipe(
-    take(1),
-    mergeMap(async (via) => {
-      if (typeof via !== 'string') via.send(content.body);
-      // via RTC channel
-      else if (via.startsWith('@')) await matrix.sendToDevice(type, { [via]: { '*': content } });
-      // via toDevice message
-      else await matrix.sendEvent(via, type, content, ''); // via room
-      // this returned value is just for notification, and shouldn't be relayed on;
-      // all functionality is provided as side effects of the subscription
-      return typeof via !== 'string' ? via.label : via;
+        );
+      else
+        return matrix$.pipe(
+          mergeMap(async (matrix) => matrix.sendEvent(via, type, content, '')),
+          mapTo(via), // via room
+          // complete only when peer is online
+          delayWhen(constant(presence$.pipe(filter((presence) => !!presence?.payload.available)))),
+        );
+    }),
+    retryWhile(intervalFromConfig(config$), {
+      maxRetries: 3,
+      onErrors: [429, 500],
```

Comment:
Should we add `networkErrors` here as well?

Reply:
Yes, we should do that.